### PR TITLE
fix(shift): chart, note, pane and pivot table positions now move when rows or columns are inserted or deleted

### PR DESCRIPTION
**Pull request by jafin**
Implements [spec 33](docs/specs/33-sheet-listener-seam.md). Seventeen sheet-scoped features had to survive a structural edit and did it four different ways — and four of them did not react at all. This makes registering an `ISheetListener` the one thing that makes a feature edit-aware, and fixes those four as a consequence.

**This changes the file XLibur writes for an existing document that is edited structurally.** Hence `fix:` and a patch bump.

## The seam, and what it replaced

`ISheetListener` was a declared seam that two types used, reached by name from four hardcoded lines in the shifter. Everything else was open-coded.

| | before | after |
|---|---|---|
| `XLWorksheetRangeShifter.cs` | 222 lines, named nine features | **65 lines, names none** |
| types implementing the port | 2 | **11** |

`XLWorksheet.GetSheetListeners()` is now the registry, modelled on the `XLWorksheets.GetWorkbookListeners()` that already existed one level up. `XLWorksheetRangeShifter` builds one `SheetEdit` and walks it. Adding a feature that must survive an insert or a delete is one adapter plus one `yield return` — the shifter does not change.

Gate: ``grep -nE 'MergedRanges|ConditionalFormats|DataValidations|PageSetup|Sparkline|DefinedNames|Hyperlinks|CalcEngine' XLibur/Excel/XLWorksheetRangeShifter.cs`` returns nothing.

Converted, each onto the type that owns it: merged-range straddle split (a new `MergedRangeSplitListener`, because the merges are a general-purpose `XLRanges` with no collection of their own), defined names, conditional-format coverage, data-validation sqref, data-validation criteria formulas, page breaks, sparkline cleanup.

The port keeps exactly four methods and gains one argument — a `readonly struct SheetEdit` carrying `Sheet`, `Area`, `Range` and `Shift`, passed `in`.

## The four features that did not move

Measured before and after, one workbook, `InsertRowsAbove(3)` then `InsertColumnsBefore(2)`:

| Feature | Anchored at | Before | After |
|---|---|---|---|
| Chart `Position` | row 10, col 3 | **row 10** | row 13 |
| Chart `SecondPosition` | row 20, col 10 | **row 20** | row 23 |
| Note `Position` on `C10` | row 9 | **row 9** | row 12 |
| `SheetView.SplitRow` / `SplitColumn` | 5 / 4 | **5 / 4** | 8 / 6 |
| `XLPivotTable.Area` | `D10` | **`D10`** | `D13` |
| Picture `TopLeftCell` (control) | `C10` | `C13` ✅ | `C13` ✅ |

The note case was the sharpest because it was *half* right: the note moved with the misc slice while its callout box stayed put, so the callout detached from the cell it belonged to. `XLComment` already documented the mechanism — "shifting rows or columns moves the note's entry within the misc slice **without telling the note**". This is the telling.

The picture was the control, and it worked for a reason worth deleting: `XLMarker` allocated a one-cell `IXLRange` purely so the range repository would shift it, and said so in a comment. That workaround is gone (task 6) — `XLMarker` holds a `Point` now and moves through the same seam. ``grep -c 'IXLRange' XLibur/Excel/Drawings/XLMarker.cs`` returns 0.

## Assertions deliberately reversed

Five, each renamed so no test claims one thing while asserting another. Every one is named in its own commit body.

| Test | was | now |
|---|---|---|
| `A_chart_anchor_does_not_move_yet` → `A_chart_anchor_moves` | `Position.Row == 10`, `SecondPosition.Row == 20` | 13, 23 |
| `A_note_anchor_does_not_move_yet` → `A_note_anchor_moves` | note on `C13`, `Position.Row == 9` | note on `C13`, `Position.Row == 12` |
| `Freeze_panes_do_not_move_yet` → `Freeze_panes_move` | `SplitRow == 5`, `SplitColumn == 4` | 8, 6 |
| `A_pivot_area_does_not_move_yet` → `A_pivot_area_moves` | `Area.FirstPoint.Row == 10` | 13 |
| `A_delete_starting_on_the_anchor_row_leaves_it_invalid_yet` → `..._clamps_it` | threw `ArgumentOutOfRangeException` | `C4:J4` |

The first four were written in task 1 asserting the **wrong** answer on purpose, so tasks 4 and 5 had to change them and could not change them silently. The fifth is D16 below.

**No other test in the suite changed, on any task.**

## The Excel-behaviour answers, and the source of each

Excel could not be opened for this work, so nothing was guessed. The answers come from a source the repo already trusts: the **picture**, the one drawing anchor that already moved correctly. Whatever the range repository does to a one-cell range is what XLibur already treats as correct for a drawing anchor — and task 6 requires the picture to behave identically afterwards, which makes reproducing it a constraint rather than a preference. Measured case by case on the unmodified tree, a two-cell picture anchored `C4:J20`:

| Edit | Result | |
|---|---|---|
| insert 3 rows at row 1 (above) | `C7:J23` | both corners move |
| insert 3 rows at row 10 (inside) | `C4:J23` | it grows |
| insert 3 rows at row 30 (below) | `C4:J20` | unchanged |
| insert 2 cols at col 1 | `E4:L20` | both corners move |
| insert 2 cols at col 5 (inside) | `C4:L20` | it grows |
| delete rows 1–2 (above) | `C2:J18` | both corners move |
| delete rows 5–6 (inside) | `C4:J18` | it shrinks |
| delete rows 1–25 (covering it) | `C1:J1` | **clamped, not deleted** |
| partial insert at `B1:B5` | `C4:J20` | outside its columns |

`GridShift` reproduces all nine and is `XLRangeShiftHelper`'s transform reduced to the integers. Those cases are now `PictureAnchorShiftTests`, written against the *old* mechanism, so task 6 moving the picture onto `GridShift` is a proof rather than a rewrite.

Settled from the format rather than from behaviour:

- **A `MoveWithCells` chart's anchor transforms identically**, and its `SecondPosition` is not written at all — `xdr:oneCellAnchor` has a `from` marker and a fixed `xdr:ext` (ECMA-376 §20.5.2.31/§20.5.2.33), and `IXLChart` documents `SecondPosition` as used only under `MoveAndSizeWithCells`.
- **An absolutely anchored drawing does not move.** `xdr:absoluteAnchor` carries `xdr:pos` in EMU with no cell reference (§20.5.2.1). The adapter skips it.
- **Chart markers are 0-based, note positions are 1-based — on the same type.** `ChartWriter` writes `Position.Row` verbatim into `xdr:row`, which §20.5.2.32 defines as a 0-based index; the VML writer emits `Position.Row - 1` and indexes `Worksheet.Row(Position.Row)` directly. Each kind declares its own base rather than the adapter assuming one.
- **A pane's split is a count, not an address.** Deleting every line above it leaves 0, and `SheetViewWriter` already removes the `pane` element outright when both counts are 0 — so zero is not a degenerate case needing a guard, it is already "no pane".

**One question could not be settled and is flagged rather than invented:** *does Excel allow a row to be inserted inside a pivot table's body, and what happens?* Excel refuses; XLibur has no such guard. What XLibur now does is well-defined and consistent with every other rectangle — the area's leading corner stays, the trailing corner moves, the area grows. Whether XLibur should instead *refuse* the edit is a separate question this spec does not answer.

## Found, not fixed

Three defects recorded rather than folded in. **D15 and D17 are live.**

- **D15 🔴 live — a hyperlink detaches from its own cell when the edited range is more than one line tall.** The area handed to listeners is the *whole* range extended by `Shift - 1`, while the cells move by `Shift`. `ws.Range("A1:A5").InsertRowsAbove(3)` leaves the content on `A8` and the hyperlink on `A12`. The two agree only when the edited range is one line tall, which is what every pre-existing test used. **Preserved deliberately** — tasks 2 and 3 had to change nothing — and pinned by `SheetEditAreaTests` so a later spec must reverse it on purpose. Fixing it means making the shifter's area agree with `XLRangeInsertHelper`'s `insertedRange`, which changes output and belongs in its own spec.
- **D16 🟡 — a delete starting exactly on a range's leading edge leaves its first address invalid** rather than clamping it, so reading it throws while every neighbouring case clamps. **Fixed for drawing anchors** by moving them onto `GridShift`; `XLRangeShiftHelper` itself is untouched, so an ordinary stored range still behaves the old way — "destroyed" may well be right for a range, and that is a separate question.
- **D17 🔴 live — a note states its anchoring mode twice and the two disagree on every note XLibur creates.** `XLComment.Initialize` sets `Anchor = MoveAndSizeWithCells`; the inherited `DefaultCommentStyle` sets `Style.Properties.Positioning = Absolute`; the VML writer reads the latter. So XLibur tells Excel every note it writes is absolutely positioned while its own object model says the opposite. The adapter gates on `Anchor` — **gating on `Positioning` would have made this fix a silent no-op for every note in existence.** Which should drive the file needs Excel.

## Performance

Spec 05 measured the range-shift pass at 8% of a 1,000-insert workload and formula shifting at 68%. This restructures the 8% and must not move the total. `dotnet run -c Release --framework net10.0 --project XLibur.Benchmarks -- profile structural`, **medians of three runs per side** — the machine has ~40% run-to-run timing variance, so single runs prove nothing.

| Probe | base ms | branch ms | Δ | base MB | branch MB | Δ |
|---|---|---|---|---|---|---|
| 1,000 inserts, 0 ranges, 0 formulas | 540 | 519 | −3.9% | 387.3 | 394.2 | +1.8% |
| 1,000 inserts, 1,000 ranges below | 692 | 664 | −4.0% | 725.0 | 728.0 | +0.4% |
| 1,000 inserts, 1,000 ranges above (no-ops) | 517 | 430 | −16.8% | 405.2 | 362.9 | −10.4% |
| 1,000 inserts, 1,000 formulas below | 913 | 905 | −0.9% | 1211.5 | 1212.9 | +0.1% |
| 1,000 inserts, 1,000 formulas above | 337 | 370 | +9.8% | 369.5 | 399.0 | +8.0% |
| **1,000 inserts, both below** | **1295** | **1283** | **−0.9%** | **1712.9** | **1712.2** | **−0.0%** |
| 1 batch insert of 1,000, both below | 3 | 3 | 0 | 2.8 | 2.8 | 0 |

**The gate caught a regression in this branch's own work.** `DrawingAnchorListener` enumerated notes with `GetCells(c => c.HasComment)`, which materialises an `XLCell` per used cell and ran on every edit whether the sheet had notes or not — a million predicate calls on the 1,000-insert probe. "1,000 formulas above" went from 337 ms / 369 MB to **450 ms / 454 MB**. It now walks the misc slice, where notes live, and returns on one branch when that slice is empty. That is what the perf task is for; caching the listener list would have hidden it rather than fixed it.

**No listener-list cache.** The spec said to cache only if an allocation regression actually showed up. After the fix above, none does.

The one apparent regression left, "1,000 formulas above" at +9.8%/+8.0%, is not a signal: it is a no-op probe whose *base* runs spread 223–370 MB, so the branch's steady 399 MB sits inside the baseline's own spread.

## Code review

Run at `high` over the whole branch. It confirmed the refactor is behaviour-preserving for the nine pre-existing passes — order, sheet guards, `CoverageArea` identical to the old `AffectedArea`, the `shift == 0` early return unreachable — and found **three defects in what this branch *added***. All three were reproduced before being fixed and all three now have tests.

1. **A note's callout still detached from its cell at the boundary — medium, and the one worth reading.** Task 4 transformed the note's anchor the way a chart's is transformed. That is wrong for a note: a chart hangs off the grid and nothing else, but a note is bound to a cell and its callout sits at an offset from it, so an anchor computing its own transform straddles the edit differently from its cell. A note on `A10` has `Position.Row == 9`; `InsertRowsAbove(3)` **at row 10** moved the cell to `A13` and left the callout at row 9 — four rows adrift, which is the exact defect this spec exists to fix, one boundary further along. **Every task-4 test passed over it**, because they all inserted clear *above* the note. The anchor now takes the cell's displacement instead. Pinned by five boundary cases plus a partial-insert case.
2. **`XLPivotTable.TargetCell` could be driven off the top of the grid — medium.** Report filters sit above the area and `TargetCell` shifts *up* by their height, so a delete clamping the area near row 1 computed a target above row 1. `Point` stores its row 0-based in an unsigned field, so it **wrapped rather than threw** — `TargetCell.Address` read `#REF!`, handed out through the public API with no error. The area now stops far enough down for the filters to fit. Unreachable before this branch, because `Area` never moved.
3. **The pane's scroll anchor was left behind by its own split — low.** `SheetViewWriter` writes `PaneTopLeftCellAddress` verbatim as `pane/@topLeftCell` when set, so moving the split without it wrote an anchor *inside* the frozen band (`ySplit="8"` against `topLeftCell="A6"`). It moves too now, by `MoveIndex` rather than `MoveCount`, because it is an address and not a count.

Two of the three share a shape worth expecting in any spec that wakes a dormant value: **the value now moves, and something derived from it was not ready for that.**

## CI found a fourth — `XLibur.Report` was compensating for one of these defects

`PivotRewriter.MovePivotTables` existed for exactly one reason, and its own remarks said so: *"a pivot table's own position is a plain rectangle too, so a pivot below a bound range stays where the template put it while the rows it sat below multiply underneath it. It is moved with them."* That is feature 17 from the table above, worked around one layer up in a different package.

With `XLPivotTable` now reacting to the row inserts expansion performs, **both movers applied the same delta**: a target at `D8` under a net two-row expansion landed on row 12 instead of 10.

Deleted rather than disabled — spec 33's point is that a feature reacts once, through one seam, and leaving a second mover beside it is exactly the "notified twice from two layers" shape the spec flags for sparklines. Verified redundant rather than assumed: with it removed, every other pivot test passes unchanged, including the two that assert a pivot does **not** move. **No user-visible change and no changelog entry** — the net output is what it always was, produced by one mechanism instead of two.

This also adds a **sixth** deliberately reversed assertion: `APivotTableDoesNotMoveWhenRowsAreInsertedAboveIt` → `APivotTableMovesWhenRowsAreInsertedAboveIt`, `TargetCell` row 1 → 5. Another characterization test pinning the old wrong behaviour — it existed to record *why* the rewriter was necessary.

**How this reached CI, stated plainly:** I ran only `XLibur.Tests` before opening, and the repo has four test projects. The spec and its dispatch brief both named only that one. The lesson is in the spec's Results: **a defect that survives long enough acquires workarounds, and they are not all in the file the spec points at.** This spec found one in shipped library code — `XLMarker`'s smuggled range, with a whole task devoted to deleting it — and made it the evidence for the entire argument. It had a sibling in another package nobody thought to look for.

## Two corrections to the spec

- **Acceptance criterion 2 is arithmetically unreachable.** It demands "at least 12" adapter types; the spec's own design section lists **11**, and 11 is what landed. Its other stated gate — the file count from `grep -rl 'ISheetListener'` — does return 12, but only because it matches the interface's own declaration file, which is not an adapter. This is the same arithmetic error spec 26's criterion 8 made, and it gets the same answer: **reporting an unreachable criterion beats splitting a type to pad the count.** Nothing was divided in two to reach a number, no documentation was deleted, no method body was copied.
- **The enumeration in the spec's design section cannot be built as written.** It yields `DataValidations` at position 5 for sqref coverage and `sheet.DataValidations` at position 6 for criteria formulas, as though the two positions could do different work. They cannot: a type implements an interface *once*, so both positions run both passes and coverage would shift **twice per edit**. Any implementation faithful to the text is wrong. What landed yields **one `XLDataValidations` per sheet**, doing sqref-if-mine then criteria — which is also the stronger encoding, since the criteria pass *must* follow the coverage pass and a method body holds that where two `yield return`s a later edit could reorder does not.

## For reviewers

- **Merge [#413](https://github.com/XLibur/XLibur/pull/413) (spec 29) before this branch.** Both carry a `docs/specs` sync from the shared source folder, so each includes the other's in-flight documentation. The conflicts are trivial and textual — **resolve by keeping both edits**. No source file is shared.
- **This does not unblock spec 34.** 34 waits on spec 27 (font conformance), not on this. The chain is 27→34; 26→33 is the one that just completed.

## Verification

- **All four test projects** green on **net8.0 and net10.0** — 29,542 tests, 0 failures: `XLibur.Tests` (28,444), `XLibur.Report.Tests` (962), `XLibur.Fonts.SixLabors.Tests` (62), `XLibur.Fonts.SkiaSharp.Tests` (74).
- No public API change — `PublicAPI.Unshipped.txt` untouched.
- Two `CHANGELOG.md` entries under `## Unreleased` → `### Fixed`, one for tasks 4 and 6, one for task 5.
- Documentation updated in the same PR: the spec's `## Results`, the tasklist boxes with commit shas, the README status cell, the dependency graph node.

## Summary by CodeRabbit

* **Enhancements**
  * Row and column insertions or deletions now consistently reposition charts, pictures, notes, pivot tables, frozen panes, page breaks, hyperlinks, validations, and named ranges.
  * Drawing anchors resize or clamp safely when edits overlap them, while absolute and free-floating drawings remain fixed.
  * Pivot tables preserve valid placement and report-filter spacing after structural edits.
  * Frozen-pane splits are adjusted automatically and removed when their entire region is deleted.

* **Bug Fixes**
  * Prevented invalid anchors and related worksheet references after structural changes.

### CODE REVIEW

File: XLibur/Excel/Drawings/DrawingAnchorListener.cs [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+            if (cellCross < editFirstCross || cellCross > editLastCross)
+                continue;
+            if (axis.IndexOf(cell) < movedFrom)
+                continue;
+
+            // 1-based, and moved by the cell's displacement, so the callout keeps its offset.
+            if (axis.ShiftsRows)
+                note.Position.SetRow(note.Position.Row + edit.Shift);
+            else
+                note.Position.SetColumn(note.Position.Column + edit.Shift);
```
[the remark this answers is not in the document]
Comment (jafin) [reply XLibur/Excel/Drawings/DrawingAnchorListener.cs]: Confirmed and fixed in `ec978058` — and it is worse than Minor: the workbook cannot be saved at all.

Reproduced before fixing. Note on `A6` (callout anchored row 5), `ws.Rows(1, 5).Delete()`:

```
before: pos.Row=5
after:  note on row 1, pos.Row=0
save:   THREW ArgumentOutOfRangeException: Row number must be between 1 and 1048576
```

So `SaveAs` fails outright rather than writing a bad coordinate.

Worth naming the cause, because it was a regression in this branch's own earlier review fix. That fix changed a note's anchor from *transforming itself* to *taking its cell's displacement*, and in doing so lost a floor the previous form had for free — `GridShift.MoveIndex` clamps at the edit's leading line; plain addition does not. A callout sits *above* its cell, so it runs out of grid before the cell does.

Floored at line 1 on both axes. Not an arbitrary clamp: `XLComment.Initialize` already makes exactly this concession for a note created on row 1 (`if (previousRowNumber > 1) previousRowNumber--`), so the box shares its note's line at the top of the sheet.

Two regression tests, and **each asserts the save, not just the coordinate** — the anchor is only wrong if it reaches the file, and a test stopping at `Position.Row == 1` would not have caught the original.

File: docs/specs/33-sheet-listener-seam.md [hunk 2]
```diff
@@ -1100,3 +1109,319 @@ git commit -m 'docs(specs): record structural-edit numbers and the Excel anchor
...
+| 6 — delete the `XLMarker` workaround | `d2e25f3a` |
+| 7 — structural-edit cost | `335a8e97` |
+| code-review fixes (3 defects in this spec's own new work) | `6d27644a` |
+| CI fix — `PivotRewriter`'s compensating mover deleted | `7ebff4c6` |
+
+All four test projects green on net8.0 and net10.0 — 29,542 tests, 0 failures: `XLibur.Tests`
+(28,444), `XLibur.Report.Tests` (962), `XLibur.Fonts.SixLabors.Tests` (62),
+`XLibur.Fonts.SkiaSharp.Tests` (74). Five assertions were deliberately
+reversed and their tests renamed; each is named in its commit body and listed below. No other test in
+the suite changed, on any task.
```
[the remark this answers is not in the document]
Comment (jafin) [reply docs/specs/33-sheet-listener-seam.md]: Fixed in `ec978058`. Six is correct — the sixth is the `XLibur.Report` pivot assertion CI found after the PR opened, and this paragraph was written before that. Test counts refreshed to 29,550 in the same edit, since the merge with #413 moved them too.

File: docs/specs/33-sheet-listener-seam.md [hunk 3]
```diff
@@ -1100,3 +1109,319 @@ git commit -m 'docs(specs): record structural-edit numbers and the Excel anchor
...
+  object yielded at position 5 and the object yielded at position 6 have the same four methods and
+  would run both passes both times. Coverage would be shifted twice per edit. Any implementation
+  faithful to the text is wrong.
+
+  **What landed instead:** the enumeration yields **one `XLDataValidations` per sheet**, and that
+  single listener does sqref-if-mine and then criteria. This is not a workaround for the interface's
+  shape — it is the stronger encoding of the requirement. The shifter's own comment says the criteria
+  pass *must* follow the coverage pass, because coverage deletes a rule whose area transforms to
+  nothing and the formula pass must not then rewrite a rule that is gone. "This method does A then B"
+  holds that ordering firmly; two `yield return`s a later edit could reorder does not.
```
[the remark this answers is not in the document]
Comment (jafin) [reply docs/specs/33-sheet-listener-seam.md:1291]: Fixed in `ec978058` — corrected the example rather than labelling it superseded, because the copy-and-run hazard you name is the real risk and a label does not remove it.

The original is kept in a comment directly above with the reason beside it, so the correction stays visible to anyone re-deriving the spec:

```csharp
// CORRECTED after implementation. This originally read:
//     yield return DataValidations;
//     foreach (var sheet in Workbook.WorksheetsInternal)
//         yield return sheet.DataValidations;   // criteria formulas, workbook-wide
// which cannot be built: a type implements the interface once, so the edited sheet's
// collection would be yielded twice and run BOTH passes both times, shifting its coverage
// twice per edit. One listener per sheet, doing sqref-if-mine then criteria in the order
// the passes require. See Results, "Two corrections to the spec".
```

This is what the Results section already asked for — *"anyone re-deriving this spec should fix the design section rather than re-discover this"* — so thanks for catching that I had not actually done it.

File: docs/specs/TASKLIST-architecture-deepening-2.md [hunk 4]
```diff
@@ -38,9 +38,9 @@ the test surface; where it has two, nothing sits at the seam to assert they agre
 | [26](26-grid-axis.md) | Give the grid one axis | L | — | ✅ **Merged** (#409; see Results) |
 | [27](27-font-conformance-suite.md) | One font conformance module | S–M | — | ⬜ Ready |
 | [28](28-single-style-decoder.md) | One OOXML style decoder | M | — | ✅ **Merged** (#411; see Results) |
-| [29](29-write-path-resolvers.md) | One resolver per emitted element | M | — | ✅ Done |
+| [29](29-write-path-resolvers.md) | One resolver per emitted element | M | — | ✅ **Merged** (#413; see Results) |
 | [30](30-array-application-seam.md) | Array application gets an interface | S–M | — | ⬜ Ready |
-| [31](31-worksheet-element-writers.md) | Worksheet element writers get one interface | M–L | **29** | ⬜ Blocked |
+| [31](31-worksheet-element-writers.md) | Worksheet element writers get one interface | M–L | 29 ✅ merged | ⬜ **Ready** |
```
[the remark this answers is not in the document]
Comment (jafin) [reply docs/specs/TASKLIST-architecture-deepening-2.md:43]: Fixed in `ec978058`. Both were left behind by the merge with #413, which landed while this branch was open:

- Spec 29's section heading said `✅ Done (2026-08-27)` → now `✅ Merged (#413, 8d2acfc7, 2026-08-27)`.
- Spec 31's heading said `⬜ Blocked on 29` → now `⬜ Ready (29 merged as #413)`, and its node in the dependency graph moved from `blocked` to `ready`. 29 landing is precisely what unblocks 31, so the graph was contradicting its own edge.

Also gave spec 33's graph node its PR number while there, for consistency with 29's.
